cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,238 @@
+
+	rfc1035       = "[a-z]([-a-z0-9]*[a-z0-9])?"
+	projectRgxStr = "[a-z]([-.:a-z0-9]*[a-z0-9])?"
+
+	metadataKeyWindowsStartupScriptURL       = "windows-startup-script-url"
+	metadataKeyWindowsStartupScriptURLBackup = "windows-startup-script-url-backup"
+
+	versionWindows2008r2 = "windows-2008r2"
+	versionWindows2012r2 = "windows-2012r2"
+
+	upgradeIntroductionTemplate = "The following resources will be created/accessed during the upgrade. " +
+		"Please keep track of their names if manual cleanup and/or rollback are necessary.\n" +
+		"All resources are in project '%v', zone '%v'.\n" +
+		"1. Instance: %v\n" +
+		"2. Disk for install media: %v\n" +
+		"3. Snapshot for original OS disk: %v\n" +
+		"4. Original OS disk: %v\n" +
+		"   - Device name of the attachment: %v\n" +
+		"   - AutoDelete of the attachment: %v\n" +
+		"5. New OS disk: %v\n" +
+		"6. Machine image: %v\n" +
+		"7. Original startup script url '%v': %v\n" +
+		"\n"
+
+	guideTemplate = "When upgrading succeeded but cleanup failed, please manually cleanup by following steps:\n" +
+		"1. Delete 'windows-startup-script-url' from the instance's metadata if there isn't an original value. " +
+		"If there is an original value, restore it. The original value is backed up as metadata 'windows-startup-script-url-backup'.\n" +
+		"2. Detach the install media disk from the instance and delete it.\n" +
+		"\n" +
+		"When upgrading failed but you didn't enable auto-rollback, or auto-rollback failed, or " +
+		"upgrading succeeded but you still need to rollback for any reason, " +
+		"please manually rollback by following steps:\n" +
+		"1. Detach the new OS disk from the instance and delete it.\n" +
+		"2. Attach the old OS disk as boot disk.\n" +
+		"3. Detach the install media disk from the instance and delete it.\n" +
+		"4. Delete 'windows-startup-script-url' from the instance's metadata if there isn't an original value. " +
+		"If there is an original value, restore it. The original value is backed up as metadata 'windows-startup-script-url-backup'.\n" +
+		"\n" +
+		"Once you verified the upgrading succeeded and decided to never rollback, you can:\n" +
+		"1. Delete the original OS disk.\n" +
+		"2. Delete the machine image.\n" +
+		"3. Delete the snapshot.\n" +
+		"\n"
+)
+
+var (
+	supportedSourceOSVersions = map[string]string{versionWindows2008r2: versionWindows2012r2}
+	supportedTargetOSVersions = reverseMap(supportedSourceOSVersions)
+
+	upgradeScriptName        = map[string]string{versionWindows2008r2: "upgrade_script_2008r2_to_2012r2.ps1"}
+	upgradeWorkflowPath      = map[string]string{versionWindows2008r2: "daisy_workflows/windows_upgrade/windows_upgrade_2008r2_to_2012r2.wf.json"}
+	retryUpgradeWorkflowPath = map[string]string{versionWindows2008r2: "daisy_workflows/windows_upgrade/windows_upgrade_2008r2_to_2012r2_retry.wf.json"}
+
+	expectedCurrentLicense = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2008-r2-dc"}
```

Comment:
What about BYOL licenses?

Reply:
It's not supported, because they didn't bring their 2012r2 license.

Reply:
I thought if a customer has windows 2008 BYOL, that would work for 2012? Velostrata is reusing 2008 license for newer Windows versions. Is that only for Pay as you go?

Reply:
BYOL doesn't work. I confirmed with Andrew at the very beginning.